### ISSUE TITLE
games-emulation/dosbox-x-0.83.25 segfault with media-libs/openglide-0.09_rc9_p20191120

### ISSUE THREAD

**Issue by bell07** [Tatsh/tatsh-overlay#140]
I play TombRaider 1 with dosbox-x because of openglide support.

Since 0.83.25 I get segmentation fault if the game try to chainge into glide mode.
dosbox-x ebuild down-bumped and downgraded to 0.83.24 it works again.

I opened already upstream issue at https://github.com/joncampbell123/dosbox-x/issues/3465 but unsure if the issue is ebuild related or dosbox-x related

**Comment by Tatsh** [comment]
You will need to debug this. I don't really blame my ebuild because this worked in a prior version. But you can use the Gentoo-only splitdebug feature to help with debugging.

Merge with `FEATURES=splitdebug` set then run dosbox-x in GDB. `gdb --args dosbox-x ...`. `c` to make the app start, then do what is necessary to cause the crash. `bt` to show a backtrace. Copy that backtrace to the dosbox-x issue so they can have a look.

**Comment by bell07** [comment]
Thank you for the debugging hints. I posted my analysis into dosbox-x issue. Seems to be related to fluidsynth compatibility. Previously the fluidsynth MIDI device was not found although fluidsynth is installed. Now MIDI device fluidsynth is found, but causes the crash in OpenGlide SDL-1 lib on initialization.
By the way, the question popped up why the ebuild uses SDL-1 instead of SDL-2?
I tried to modify the ebuild byself and got dosbox-x compiled with SDL-2, but this did not solved my issue :-(

**Comment by Tatsh** [comment]
> By the way, the question popped up why the ebuild uses SDL-1 instead of SDL-2?

Back when this ebuild was originally made, either only SDL1 was supported or there was an issue with SDL2.

**Comment by bell07** [comment]
I tried the new SDL2 ebuild. The issue still apears unchanged. Seems fluidsynth breaks SDL1 in OpenGlide in any way. Used the new ebuild renamed to 0.83.24, the old version, that was working with SDL1, does not work with SDL2 anymore. Same with 0.83.23. Fluidsynth seems to be detected with new ebuild version. Previously in SDL1 ebuild fluidsynth was not found

**Comment by Tatsh** [comment]
Well it was not intended to fix your issue but I wanted to remove the extra code that builds a custom version of SDL1 when SDL2 seems to work otherwise now. Looks like this really is an upstream issue? In that case, I'd like to close this ticket.

**Comment by bell07** [comment]
Since the root reason is still not found, I do not know if it is related to upstream or to gentoo ebuild. Does OpenGlide work for you?

**Comment by Tatsh** [comment]
I can try but I assume you want me to try a specific version of Tomb Raider?

**Comment by bell07** [comment]
No. Same issue with any Glide applications. You can try it with demo at http://falconfly.3dfx.pl/coolstuff.htm

**Comment by Tatsh** [comment]
TLDR: Try disabling SDL: `env USE=-sdl emerge openglide`.

Got the same kind of error with the Nature demo with OpenGlide as it normally installs. The backtrace shows it's in SDL 1's code.

```
(ins)(gdb)
#0  0x00007fa03052026c in SDL_ListModes (format=format@entry=0x7ffd16356230, flags=flags@entry=2)
    at /var/tmp/portage/media-libs/libsdl-1.2.15_p20210224/work/SDL-1.2-99d7f1d1c5492f0fb3c799255042ca7a3f4a5de4/src/video/SDL_video.c:353
(ins)(gdb) list
348             SDL_VideoDevice *video = current_video;
349             SDL_VideoDevice *this  = current_video;
350             SDL_Rect **modes;
351
352             modes = NULL;
353             if ( SDL_VideoSurface ) {
354                     if ( format == NULL ) {
355                             format = SDL_VideoSurface->format;
356                     }
357                     modes = video->ListModes(this, format, flags);
```

I wrote some sample code to call `SDL_ListModes` that according to SDL 1 documentation should work `gcc -o test test.c -lSDL`:

```c
#include <stdio.h>
#include <SDL/SDL.h>

int main(int argc, char *argv[]) {
    SDL_Rect **modes;
    modes = SDL_ListModes(NULL, SDL_FULLSCREEN|SDL_HWSURFACE);
if(modes == (SDL_Rect **)-1){
  printf("All resolutions available.\n");
}
else{
  /* Print valid modes */
  printf("Available Modes\n");
    int i;
  for(i=0;modes[i];++i)
    printf("  %d x %d\n", modes[i]->w, modes[i]->h);
}
    return 0;
}
```

It gets the same exact crash as OpenGlide.

On line 353, `SDL_VideoSurface` is NULL. In the generated assembly, there's a bad mov instruction. GDB says there's no `SDL_VideoSurface` variable at all.

```
Dump of assembler code for function SDL_ListModes:
   0x00007fa030520260 <+0>:     mov    r8,rdi
   0x00007fa030520263 <+3>:     mov    rdi,QWORD PTR [rip+0x37dce]        # 0x7fa030558038 <current_video>
   0x00007fa03052026a <+10>:    mov    edx,esi
=> 0x00007fa03052026c <+12>:    mov    rax,QWORD PTR [rdi+0x250]
...
(ins)(gdb) p SDL_VideoSurface
No symbol "SDL_VideoSurface" in current context.
```

`QWORD PTR [$rdi+0x250]` is not accessible. `$rdi` is `0` here so it's trying to access `0x250`, which is not going to work:

```
(ins)(gdb) x/5i $rdi+0x250
   0x250:       Cannot access memory at address 0x250
```

I think this comment in SDL1's code is relevant. It should say that any function that tries to get `SDL_VideoSurface` will segfault.

```c
        /*
         * rcg11292000
         * If you try to set an SDL_OPENGL surface, and fail to find a
         * matching  visual, then the next call to SDL_SetVideoMode()
         * will segfault, since  we no longer point to a dummy surface,
         * but rather NULL.
         * Sam 11/29/00
         * WARNING, we need to make sure that the previous mode hasn't
         * already been freed by the video driver.  What do we do in
         * that case?  Should we call SDL_VideoInit() again?
         */
        SDL_VideoSurface = (mode != NULL) ? mode : prev_mode;
```

With the USE change, the Nature demo on that page runs but it takes a while to load. Once it does, it seems to work.

**Comment by Tatsh** [comment]
I have no idea why it takes so long to load. I even set `cycles = max` but it doesn't change the speed of what is happening in the OpenGlide window. I am guessing the majority of the code in that demo is actual Glide code that does non-GPU things during the initialisation.

**Comment by bell07** [comment]
I tried USE="-sdl" for openglide before. The game launches fast, but I cannot play the game because the keyboard seems to be not connected to the game.

**Comment by bell07** [comment]
Got managed 3dfx without openglide. Can you pls revert the patch https://github.com/joncampbell123/dosbox-x/commit/1624f0490e7a0a7f8fa718609483839617c564f5 in ebuild? Without them 3dfx is able to scale window.

**Comment by Tatsh** [comment]
Can you report your issue with that change to upstream?

**Comment by Tatsh** [comment]
You should submit a patch for the current OpenGlide to fix the SDL issues. It appears the fork you found uses SDL 2 which is an improvement as SDL 1 can get closer to removal.

**Comment by bell07** [comment]
I am not a developer, so I cannot do the patching. But I added an comment to an existing issue at https://github.com/voyageur/openglide/issues/20#issuecomment-1134675862. Hope any developer pick it up and update the OpenGlide.

**Comment by Tatsh** [comment]
Closing because this seems to be not an issue with anything here.
